### PR TITLE
Added modifier for ongoing raid, raid victory and raid defeat.

### PR DESCRIPTION
**Pull request by Zetabite**
Also. Atmosfera debug log now tries to use String.format() where possible.

### CODE REVIEW

File: src/main/java/dev/hephaestus/atmosfera/world/context/Sphere.java [hunk 1]
```diff
@@ -126,6 +132,39 @@ public void update(ClientPlayerEntity player) {
...
+            this.isDefeatedInRaid = false;
+            this.isVictoriousInRaid = false;
+            this.isInWitherFight = false;
+            this.isInEnderDragonFight = false;
+            if (!bossBars.isEmpty()) {
+                Iterator var1 = bossBars.values().iterator();
+
+                while(var1.hasNext()) {
+                    BossBar bossBar = (BossBar)var1.next();
+                    MutableText bossBarText = bossBar.getName().copy();
```
Comment (Haven-King) [inline src/main/java/dev/hephaestus/atmosfera/world/context/Sphere.java:148]: Making a copy here is somewhat wasteful and not really necessary.

File: src/main/java/dev/hephaestus/atmosfera/world/context/Sphere.java [hunk 2]
```diff
@@ -126,6 +132,39 @@ public void update(ClientPlayerEntity player) {
...
+                    if(bossBarText.toString().contains("'event.minecraft.raid'")) {
```
Comment (Haven-King) [inline src/main/java/dev/hephaestus/atmosfera/world/context/Sphere.java:149]: This logic will end up moving to the actual condition, but calling `toString` isn't necessary here either. I'd do something like
```java
String value = bossBarText instanceof TranslatableText translatable ? translatable.getKey() : bossBarText.asString();
```

**Review by Haven-King — CHANGES_REQUESTED** [review]
There are some things I'd like to see changed before this is merged, but it's a good start 🙂. Let me know if you have any questions!

File: src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.Iterator;
+import java.util.regex.Pattern;
+
+public record BossBarCondition(String text, boolean use_regex) implements AtmosphericSoundModifier, AtmosphericSoundModifier.Factory {
+
+    @Override
+    public float getModifier(EnvironmentContext context) {
+        if(!(text instanceof String)) {
+            return 0;
+        }
```
Comment (Haven-King) [inline src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java:21]: A null check would be more appropriate here, but even so, `text` should not be null at this point. That's the kind of thing you'd want to check in the deserialization method, `create(JsonObject)`.

File: src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            Iterator var1 = bossBars.iterator();
+            while(var1.hasNext()) {
+                BossBar bossBar = (BossBar) var1.next();
+                String value = bossBar.getName() instanceof TranslatableText translatable ? translatable.getKey() : bossBar.getName().asString();
+
+                if(!use_regex && value.contains(text)) {
+                    return 1;
+                } else if(use_regex && Pattern.matches(text, value)) {
+                    return 1;
+                }
```
Comment (Haven-King) [inline src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java:35]: Once we have the `Collection<String>` mentioned below, we can shorten this method quite a bit:
```java
if (this.isRegex) {
  for (String string : context.getBossBars()) {
    if (Pattern.matches(this.text, string)) return 1;
  }
} else if (context.getBossBars().contains(this.text)) {
  return 1;
}

return 0;
```

File: src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import net.minecraft.client.gui.hud.ClientBossBar;
+import net.minecraft.entity.boss.BossBar;
+import net.minecraft.text.TranslatableText;
+import net.minecraft.world.World;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.regex.Pattern;
+
+public record BossBarCondition(String text, boolean use_regex) implements AtmosphericSoundModifier, AtmosphericSoundModifier.Factory {
```
Comment (Haven-King) [inline src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java:15]: By general Java naming conventions, camelCase should be used over snake_case. I would therefore probably name `use_regex` `isRegex`.

**Review by Haven-King — CHANGES_REQUESTED** [review]

File: src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java (hunk 5 above)
Comment (Zetabite) [reply src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java:15]: its a bad habit from lua, yeah, will do

File: src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        if(!(text instanceof String)) {
+            return 0;
+        }
+
+        Collection<ClientBossBar> bossBars = context.getBossBars();
+
+        if (!bossBars.isEmpty()) {
+            Iterator var1 = bossBars.iterator();
+            while(var1.hasNext()) {
+                BossBar bossBar = (BossBar) var1.next();
```
[the remark this answers is not in the document]
Comment (Zetabite) [reply src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java:28]: nope. was just tunnel thinking

File: src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    public float getModifier(EnvironmentContext context) {
+        if(!(text instanceof String)) {
+            return 0;
+        }
+
+        Collection<ClientBossBar> bossBars = context.getBossBars();
+
+        if (!bossBars.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (Zetabite) [reply src/main/java/dev/hephaestus/atmosfera/client/sound/modifiers/implementations/BossBarCondition.java:25]: i have no complete idea how it worked, so this is actually the first time i here that, will remember

**Review by Haven-King — CHANGES_REQUESTED** [review]
I'm fine with re-adding the submerged modifier, but I'd like it to be more open to other fluids and tags similar to the blocks modifier, instead of being limited to just water.

The boss bar stuff looks good 🙂

**Comment by Zetabite** [comment]
Ight, i remove it for now, not to hard to add. I will take care of the definitions in the example pack too

**Review by Haven-King — APPROVED** [review]

**Pull request merged**